File: math/WinRT/Vector2.cpp
// Copyright (c) Microsoft Corporation.  All rights reserved

#include "pch.h"
#include "MathTypeConversion.h"


class Vector2Statics : public Microsoft::WRL::ActivationFactory<Midl::IVector2HelperStatics>
{
    InspectableClassStatic(RuntimeClass_Microsoft_DirectX_Math_Vector2Helper, BaseTrust);

public:
    IFACEMETHODIMP FromValues(float x, float y, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2(x, y);
        return S_OK;
    }

    IFACEMETHODIMP FromValue(float value, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2(value);
        return S_OK;
    }

    IFACEMETHODIMP FromPoint(ABI::Windows::Foundation::Point value, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2(value.X, value.Y);
        return S_OK;
    }

    IFACEMETHODIMP FromSize(ABI::Windows::Foundation::Size value, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2(value.Width, value.Height);
        return S_OK;
    }

    IFACEMETHODIMP ToPoint(Midl::Vector2 value, _Out_ ABI::Windows::Foundation::Point* ret)
    {
        ret->X = value.X;
        ret->Y = value.Y;
        return S_OK;
    }

    IFACEMETHODIMP ToSize(Midl::Vector2 value, _Out_ ABI::Windows::Foundation::Size* ret)
    {
        ret->Width = value.X;
        ret->Height = value.Y;
        return S_OK;
    }

    IFACEMETHODIMP get_Zero(_Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Zero();
        return S_OK;
    }

    IFACEMETHODIMP get_One(_Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::One();
        return S_OK;
    }

    IFACEMETHODIMP get_UnitX(_Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::UnitX();
        return S_OK;
    }

    IFACEMETHODIMP get_UnitY(_Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::UnitY();
        return S_OK;
    }

    IFACEMETHODIMP Length(Midl::Vector2 value, _Out_ float* ret)
    {
        *ret = asImpl(value).Length();
        return S_OK;
    }

    IFACEMETHODIMP LengthSquared(Midl::Vector2 value, _Out_ float* ret)
    {
        *ret = asImpl(value).LengthSquared();
        return S_OK;
    }

    IFACEMETHODIMP Distance(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ float* ret)
    {
        *ret = Impl::Vector2::Distance(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP DistanceSquared(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ float* ret)
    {
        *ret = Impl::Vector2::DistanceSquared(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Dot(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ float* ret)
    {
        *ret = Impl::Vector2::Dot(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Normalize(Midl::Vector2 value, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Normalize(asImpl(value));
        return S_OK;
    }

    IFACEMETHODIMP Reflect(Midl::Vector2 vector, Midl::Vector2 normal, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Reflect(asImpl(vector), asImpl(normal));
        return S_OK;
    }

    IFACEMETHODIMP Min(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Min(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Max(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Max(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Clamp(Midl::Vector2 value1, Midl::Vector2 min, Midl::Vector2 max, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Clamp(asImpl(value1), asImpl(min), asImpl(max));
        return S_OK;
    }

    IFACEMETHODIMP Lerp(Midl::Vector2 value1, Midl::Vector2 value2, float amount, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Lerp(asImpl(value1), asImpl(value2), amount);
        return S_OK;
    }

    IFACEMETHODIMP TransformByMatrix3x2(Midl::Vector2 position, Midl::Matrix3x2 matrix, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Transform(asImpl(position), asImpl(matrix));
        return S_OK;
    }

    IFACEMETHODIMP Transform(Midl::Vector2 position, Midl::Matrix4x4 matrix, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Transform(asImpl(position), asImpl(matrix));
        return S_OK;
    }

    IFACEMETHODIMP TransformNormalByMatrix3x2(Midl::Vector2 normal, Midl::Matrix3x2 matrix, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::TransformNormal(asImpl(normal), asImpl(matrix));
        return S_OK;
    }

    IFACEMETHODIMP TransformNormal(Midl::Vector2 normal, Midl::Matrix4x4 matrix, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::TransformNormal(asImpl(normal), asImpl(matrix));
        return S_OK;
    }

    IFACEMETHODIMP TransformByQuaternion(Midl::Vector2 value, Midl::Quaternion rotation, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Transform(asImpl(value), asImpl(rotation));
        return S_OK;
    }

    IFACEMETHODIMP Negate(Midl::Vector2 value, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Negate(asImpl(value));
        return S_OK;
    }

    IFACEMETHODIMP Add(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Add(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Subtract(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Subtract(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP Multiply(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Multiply(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP MultiplyByScalar(Midl::Vector2 value1, float value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Multiply(asImpl(value1), value2);
        return S_OK;
    }

    IFACEMETHODIMP Divide(Midl::Vector2 value1, Midl::Vector2 value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Divide(asImpl(value1), asImpl(value2));
        return S_OK;
    }

    IFACEMETHODIMP DivideByScalar(Midl::Vector2 value1, float value2, _Out_ Midl::Vector2* ret)
    {
        *asImpl(ret) = Impl::Vector2::Divide(asImpl(value1), value2);
        return S_OK;
    }
};


ActivatableStaticOnlyFactory(Vector2Statics);

File: numerics/DotNet/Vector2.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// Licensed under the MIT License. See LICENSE.txt in the project root for license information.

using System.Globalization;

using SM = System.Math;

namespace System.Numerics
{
    public struct Vector2 : IEquatable<Vector2>
    {
        public float X;
        public float Y;


        public static Vector2 Zero  { get { return new Vector2(); } }
        public static Vector2 One   { get { return new Vector2(1.0f, 1.0f); } }
        public static Vector2 UnitX { get { return new Vector2(1.0f, 0.0f); } }
        public static Vector2 UnitY { get { return new Vector2(0.0f, 1.0f); } }


        public Vector2(float x, float y)
        {
            X = x; 
            Y = y;
        }


        public Vector2(float value)
        {
            X = Y = value;
        }


        public float Length()
        {
            float ls = X * X + Y * Y;
            
            return (float)SM.Sqrt((double)ls);
        }


        public float LengthSquared()
        {
            return X * X + Y * Y;
        }


        public static float Distance(Vector2 value1, Vector2 value2)
        {
            float dx = value1.X - value2.X;
            float dy = value1.Y - value2.Y;
            
            float ls = dx * dx + dy * dy;

            return (float)SM.Sqrt((double)ls);
        }


        public static float DistanceSquared(Vector2 value1, Vector2 value2)
        {
            float dx = value1.X - value2.X;
            float dy = value1.Y - value2.Y;
            
            return dx * dx + dy * dy;
        }


        public static float Dot(Vector2 value1, Vector2 value2)
        {
            return value1.X * value2.X + 
                   value1.Y * value2.Y;
        }


        public static Vector2 Normalize(Vector2 value)
        {
            Vector2 ans;
            
            float ls = value.X * value.X + value.Y * value.Y;
            float invNorm = 1.0f / (float)SM.Sqrt((double)ls);

            ans.X = value.X * invNorm;
            ans.Y = value.Y * invNorm;

            return ans;
        }


        public static Vector2 Reflect(Vector2 vector, Vector2 normal)
        {
            Vector2 ans;

            float dot = vector.X * normal.X + vector.Y * normal.Y;

            ans.X = vector.X - 2.0f * dot * normal.X;
            ans.Y = vector.Y - 2.0f * dot * normal.Y;

            return ans;
        }


        public static Vector2 Min(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;
            
            ans.X = (value1.X < value2.X) ? value1.X : value2.X;
            ans.Y = (value1.Y < value2.Y) ? value1.Y : value2.Y;
            
            return ans;
        }


        public static Vector2 Max(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;
            
            ans.X = (value1.X > value2.X) ? value1.X : value2.X;
            ans.Y = (value1.Y > value2.Y) ? value1.Y : value2.Y;
            
            return ans;
        }


        public static Vector2 Clamp(Vector2 value1, Vector2 min, Vector2 max)
        {
            Vector2 result;

            // This compare order is very important!!!
            // We must follow HLSL behavior in the case user specfied min value is bigger than max value.
            float x = value1.X;
            x = (x > max.X) ? max.X : x;
            x = (x < min.X) ? min.X : x;

            float y = value1.Y;
            y = (y > max.Y) ? max.Y : y;
            y = (y < min.Y) ? min.Y : y;

            result.X = x;
            result.Y = y;

            return result;
        }


        public static Vector2 Lerp(Vector2 value1, Vector2 value2, float amount)
        {
            Vector2 ans;

            ans.X = value1.X + (value2.X - value1.X) * amount;
            ans.Y = value1.Y + (value2.Y - value1.Y) * amount;

            return ans;
        }


        public static Vector2 Transform(Vector2 position, Matrix3x2 matrix)
        {
            Vector2 result;

            result.X = position.X * matrix.M11 + position.Y * matrix.M21 + matrix.M31;
            result.Y = position.X * matrix.M12 + position.Y * matrix.M22 + matrix.M32;

            return result;
        }


        public static Vector2 Transform(Vector2 position, Matrix4x4 matrix)
        {
            Vector2 result;

            result.X = position.X * matrix.M11 + position.Y * matrix.M21 + matrix.M41;
            result.Y = position.X * matrix.M12 + position.Y * matrix.M22 + matrix.M42;

            return result;
        }


        public static Vector2 TransformNormal(Vector2 normal, Matrix3x2 matrix)
        {
            Vector2 result;

            result.X = normal.X * matrix.M11 + normal.Y * matrix.M21;
            result.Y = normal.X * matrix.M12 + normal.Y * matrix.M22;

            return result;
        }


        public static Vector2 TransformNormal(Vector2 normal, Matrix4x4 matrix)
        {
            Vector2 result;

            result.X = normal.X * matrix.M11 + normal.Y * matrix.M21;
            result.Y = normal.X * matrix.M12 + normal.Y * matrix.M22;

            return result;
        }


        public static Vector2 Transform(Vector2 value, Quaternion rotation)
        {
            Vector2 result;

            float x2 = rotation.X + rotation.X;
            float y2 = rotation.Y + rotation.Y;
            float z2 = rotation.Z + rotation.Z;

            float wz2 = rotation.W * z2;
            float xx2 = rotation.X * x2;
            float xy2 = rotation.X * y2;
            float yy2 = rotation.Y * y2;
            float zz2 = rotation.Z * z2;

            result.X = value.X * (1.0f - yy2 - zz2) + value.Y * (xy2 - wz2);
            result.Y = value.X * (xy2 + wz2) + value.Y * (1.0f - xx2 - zz2);

            return result;
        }


        public static Vector2 Negate(Vector2 value)
        {
            Vector2 ans;

            ans.X = -value.X;
            ans.Y = -value.Y;

            return ans;
        }


        public static Vector2 Add(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X + value2.X;
            ans.Y = value1.Y + value2.Y;

            return ans;
        }


        public static Vector2 Subtract(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X - value2.X;
            ans.Y = value1.Y - value2.Y;

            return ans;
        }


        public static Vector2 Multiply(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X * value2.X;
            ans.Y = value1.Y * value2.Y;

            return ans;
        }


        public static Vector2 Multiply(Vector2 value1, float value2)
        {
            Vector2 ans;

            ans.X = value1.X * value2;
            ans.Y = value1.Y * value2;

            return ans;
        }


        public static Vector2 Divide(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X / value2.X;
            ans.Y = value1.Y / value2.Y;

            return ans;
        }


        public static Vector2 Divide(Vector2 value1, float value2)
        {
            Vector2 ans;

            float invDiv = 1.0f / value2;

            ans.X = value1.X * invDiv;
            ans.Y = value1.Y * invDiv;

            return ans;
        }


        public static Vector2 operator -(Vector2 value)
        {
            Vector2 ans;

            ans.X = -value.X;
            ans.Y = -value.Y;

            return ans;
        }


        public static Vector2 operator +(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X + value2.X;
            ans.Y = value1.Y + value2.Y;

            return ans;
        }


        public static Vector2 operator -(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X - value2.X;
            ans.Y = value1.Y - value2.Y;

            return ans;
        }


        public static Vector2 operator *(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X * value2.X;
            ans.Y = value1.Y * value2.Y;

            return ans;
        }


        public static Vector2 operator *(Vector2 value1, float value2)
        {
            Vector2 ans;

            ans.X = value1.X * value2;
            ans.Y = value1.Y * value2;

            return ans;
        }


        public static Vector2 operator *(float value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value2.X * value1;
            ans.Y = value2.Y * value1;

            return ans;
        }


        public static Vector2 operator /(Vector2 value1, Vector2 value2)
        {
            Vector2 ans;

            ans.X = value1.X / value2.X;
            ans.Y = value1.Y / value2.Y;

            return ans;
        }


        public static Vector2 operator /(Vector2 value1, float value2)
        {
            Vector2 ans;

            float invDiv = 1.0f / value2;
            
            ans.X = value1.X * invDiv;
            ans.Y = value1.Y * invDiv;

            return ans;
        }


        public static bool operator ==(Vector2 value1, Vector2 value2)
        {
            return (value1.X == value2.X &&
                    value1.Y == value2.Y);
        }


        public static bool operator !=(Vector2 value1, Vector2 value2)
        {
            return (value1.X != value2.X ||
                    value1.Y != value2.Y);
        }


        public bool Equals(Vector2 other)
        {
            return X == other.X &&
                   Y == other.Y;
        }


        public override bool Equals(object obj)
        {
            if (obj is Vector2)
            {
                return Equals((Vector2)obj);
            }

            return false;
        }


        public override string ToString()
        {
            CultureInfo ci = CultureInfo.CurrentCulture;

            return String.Format(ci, "{{X:{0} Y:{1}}}", X.ToString(ci), Y.ToString(ci));
        }


        public override int GetHashCode()
        {
            return X.GetHashCode() + Y.GetHashCode();
        }

        
#if INCLUDE_WINRT_CANVAS_INTEROP

        public static unsafe implicit operator Microsoft.Graphics.Canvas.Numerics.Vector2(Vector2 value)
        {
            return *(Microsoft.Graphics.Canvas.Numerics.Vector2*)&value;
        }


        public static unsafe implicit operator Vector2(Microsoft.Graphics.Canvas.Numerics.Vector2 value)
        {
            return *(Vector2*)&value;
        }

#endif
    }
}
